manticore/ethereum.py

```diff
@@ -68,12 +92,115 @@ def report(self, state):
             output += '\t Program counter: %s\n' % pc
             output += '\t Snippet:\n'
-            output += '\n'.join(('\t\t' + x for x in self._get_src(address, pc).split('\n')))
+            output += '\n'.join(('\t ' + x for x in self._get_src(address, pc).split('\n')))
             output += '\n'
         return output
 
 
-class IntegerOverflow(Detector):
+class FilterFunctions(Plugin):
+    def __init__(self, regexp=r'.*', mutability='both', depth='both', fallback=False, include=True, **kwargs):
+        """
+            Constrain input based on function metadata. Include or avoid functions selected by the specified criteria.
+
+            Examples:
+            #Do not explore any human transactions that end up calling a constant function
+            no_human_constant = FilterFunctions(depth='human', mutability='constant', include=False)
+
+            #At human tx depth only accept synthetic check functions
+            only_tests = FilterFunctions(regexp=r'mcore_.*', depth='human', include=False)
+
+            :param regexp: a regular expresion over the name of the function '.*' will match all functions
+            :param mutability: mutable, constant or both will match functions declared in the abi to be of such class
+            :param depth: match functions in internal transactions, in human initiated transactions or in both types
+            :param fallback: if True include the fallback function. Hash will be 00000000 for it
+            :param include: if False exclude the selected functions, if True include them
+        """
+        super(FilterFunctions, self).__init__(**kwargs)
+        depth = depth.lower()
+        if depth not in ('human', 'internal', 'both'):
+            raise ValueError
+        mutability = mutability.lower()
+        if mutability not in ('mutable', 'constant', 'both'):
+            raise ValueError
+
+        #fixme better names for member variables
+        self._regexp = regexp
+        self._mutability = mutability
+        self._depth = depth
+        self._fallback = fallback
+        self._include = include
+
+    def will_open_transaction_callback(self, state, tx):
+        world = state.platform
+        tx_cnt = len(world.all_transactions)
+        # Constrain input only once per tx, per plugin
+        if state.context.get('constrained%d' % id(self), 0) != tx_cnt:
+            state.context['constrained%d' % id(self)] = tx_cnt
+
+            if self._depth == 'human' and not tx.is_human:
+                return
+            if self._depth == 'internal' and tx.is_human:
+                return
+
+            #Get metadata if any for the targe addreess of current tx
+            md = self.manticore.get_metadata(tx.address)
+            if md is None:
+                return
+            #Lets compile  the list of interesting hashes
+            selected_functions = []
+
+            for func_hsh in md.hashes:
+                if func_hsh == '00000000':
```

Comment:
should this be a `bytes` literal?

Reply:
idk. That's the hex representation of the 4 byte binary. It'll always be ascii.